python/dglke/models/ke_model.py

```diff
@@ -0,0 +1,976 @@
+                rel = rel[rel_idx]
+                tail = tail[tail_idx]
+            elif exec_mode == 'batch_rel':
+                tail_idx = idx % num_tail
+                idx = floor_divide(idx, num_tail)
+                head_idx = idx % num_head
+
+                head = head[head_idx]
+                rel = th.full((topk,), rel, dtype=rel.dtype)
+                tail = tail[tail_idx]
+            elif exec_mode == 'batch_tail':
+                rel_idx = idx % num_rel
+                idx = floor_divide(idx, num_rel)
+                head_idx = idx % num_head
+
+                head = head[head_idx]
+                rel = rel[rel_idx]
+                tail = th.full((topk,), tail, dtype=tail.dtype)
+
+            if exclude_mode == 'mask':
+                # Find exising edges
+                # It is expacted that the existing edges are much less than triples
+                # The idea is: 1) we get existing edges using g.edge_ids
+                #              2) sort edges according to source node id (O(nlog(n)), n is number of edges)
+                #              3) sort candidate triples according to cur_head (O(mlog(m)), m is number of cur_head nodes)
+                #              4) go over all candidate triples and compare with existing edges and mask them,
+                #                 as both edges and candidate triples are sorted. filtering edges out
+                #                 will take only O(n+m)
+                uid, vid, eid = g.edge_ids(head, tail, return_uv=True)
+                rid = g.edata[self._etid_field][eid]
+                mask = th.full((head.shape[0],), False, dtype=th.bool)
+
+                if len(uid) > 0:
+                    for i in range(head.shape[0]):
+                        h = head[i]
+                        r = rel[i]
+                        t = tail[i]
+
+                        h_where = uid == h
+                        t_where = vid[h_where] == t
+                        r_where = rid[h_where][t_where]
+                        if r_where.shape[0] > 0:
+                            for c_r in r_where:
+                                if c_r == r:
+                                    mask[i] = True
+                                    break
+
+                result.append((head, rel, tail, score, mask))
+            else:
+                result.append((head, rel, tail, score, None))
+
+        return result
+
+    def _topk_exclude_pos(self, score, idx, head, rel, tail, topk, exec_mode, exclude_mode):
+        """ Generate topk most relevent triplets and corresponding scores.
+
+            It takes following steps:
+
+              1) find topk elements
+              2) sort topk elements in descending order
+              3) call _exclude_pos if figure out existing edges
+        """
+        if exclude_mode == 'exclude':
+            if idx.shape[0] < topk * 4: # TODO(xiangsx): Find a better value of topk * n
```

Comment:
do we need to check this? can we always use the else branch?

Reply:
You cannot do topk when k is larger than len(score)